### ISSUE TITLE
mvcc-compliant index backfiller: filtering erroneous deletes

### ISSUE THREAD

**Issue by stevendanna** [cockroachdb/cockroach#75658]
The new mvccc-compliant index backfiller uses a "temporary index" with a delete preserving encoding to handle writes that happend during the AddSStable-based backfill.  This temporary index is then merged into the newly added index before it is validated and brought online.

The temporary index steps through `DELETE-ONLY` and `DELETE-AND-WRITE-ONLY` like any other index. However, since it is using the delete preserving encoding, the following sequence of events:

    t0: Insert of a=5 in primary index
    t1: CREATE INDEX adds temporary index mutation with DELETE-ONLY state
    t2: Delete of a=5
    t3: Insert of a=5
    t4: Temporary index mutation stepped to DELETE-AND-WRITE-ONLY

From here, the temporary index would contain a preserved delete of a=5, but not the subsequent write. Merging this index into the backfilled index, would erroneously delete a=5.

To solve for this, we wanted to persist an initial read timestamp at the start of the AddSSTable backfilling process. Any entries below that timestamp in the temporary index are ignored during the merge.

However, @dt points out that this won't work in the face of restore which will change the timestamps on restored entries in the delete preserving index.

My current line of thinking to solve this is to store _something_ in the body of the preserved delete.  Below are two options, each of which with its own potential problems.  Note that the problems haven't been verified and are just the result of thinking through the options.

- Option 1: Store a timestamp

  The filtering in Merge would then work similar to how it works now. But, which timestamp can we store?  At the point we construct the payload for the write, we don't know the actual timestamp that we will write at (for instance if get pushed). 

- Option 2: Store the table's descriptor version

  Storing the descriptor version would give us a rough "logical time" for that table that the delete wrote at. However, the  exact filtering logic gets a bit trickier. And I _think_ it would require us to force a new descriptor version as part of
  the index backfiller flow.

  Currently, during the mvcc-compliant index add process, we go through a few descriptor versions:

      version N
            new-index:  BACKFILLING
            temp-index: DELETE-ONLY
      
      version N+1
            new-index: BACKFILLING
            temp-index: DELETE-AND-WRITE-ONLY
      
      AddSSTable Backfill
      
      version N+2
            new-index: DELETE-ONLY
            temp-index: DELETE-AND-WRITE-ONLY
      
      version N+3
            new-index: DELETE-AND-WRITE-ONLY
            temp-index: DELETE-AND-WRITE-ONLY
      
      Merge

  At the point of the Merge, we would like to filter out any deletes that occured before the AddSSTable started. But, since we allow up to 2 descriptor versions in flight at a time, I believe there could be valid deletes (that need to be merged) that happen at table version N, but at a timestamp after the AddSStable backfill started. If we forced another version (M) before the AddSSTable backfill, then I believe we could filter out any deletes with version M - 2 or before, as we would know those had to come before the AddSStable backfill started.

  Also, we'll need to confirm that on restore, all descriptor versions are restored in a way that makes this scheme viable.

**Comment by Tazdaddy1** [comment]
Wyd you are fine I have a few minutes

**Comment by Tazdaddy1** [comment]
Wyd you are fine I have a few minutes

**Comment by ajwerner** [comment]
What if, in DELETE_ONLY we write the deletes to the temp index as real deletes and not as preserved deletes? I think we don't need to preserve any deletes that come at a timestamp before all nodes are using WRITE_ONLY, because we wait for that before starting the backfill.

**Comment by stevendanna** [comment]
> What if, in DELETE_ONLY we write the deletes to the temp index as real deletes and not as preserved deletes?
> I think we don't need to preserve any deletes that come at a timestamp before all nodes are using WRITE_ONLY, because we wait for that before starting the backfill.

I think this can work. Or at least, I haven't thought of a counter-example yet.  The existing schema changer tests were pretty good about finding the original bug here, so I'll code this up and see if they yield any problems.

> because we wait for that before starting the backfill.

Right, the waiting for a single version likely also allows us to avoid adding a new version in Option 2; but I like your alternative better.

**Comment by ajwerner** [comment]
> Right, the waiting for a single version likely also allows us to avoid adding a new version in Option 2; but I like your alternative better.

We always need to wait for a single version after setting `WRITE_ONLY` before backfilling.

**Comment by stevendanna** [comment]
> We always need to wait for a single version after setting WRITE_ONLY before backfilling.

Yup, I believe we currently do as we use `sc.txn` which uses `(*CollectionFactory).Txn` which does waits for a single version or no version depending on the situation.
